pkg/proxy/conntrack/cleanup.go

```diff
@@ -20,95 +20,109 @@ limitations under the License.
-	var filters []netlink.CustomConntrackFilter
-	conntrackCleanupServiceIPs := serviceUpdateResult.DeletedUDPClusterIPs
-	conntrackCleanupServiceNodePorts := sets.New[int]()
-
-	// merge newly active services gathered from endpointsUpdateResult
-	// a UDP service that changes from 0 to non-0 endpoints is newly active.
-	for _, svcPortName := range endpointsUpdateResult.NewlyActiveUDPServices {
-		if svcInfo, ok := svcPortMap[svcPortName]; ok {
-			klog.V(4).InfoS("Newly-active UDP service may have stale conntrack entries", "servicePortName", svcPortName)
-			conntrackCleanupServiceIPs.Insert(svcInfo.ClusterIP().String())
-			for _, extIP := range svcInfo.ExternalIPs() {
-				conntrackCleanupServiceIPs.Insert(extIP.String())
-			}
-			for _, lbIP := range svcInfo.LoadBalancerVIPs() {
-				conntrackCleanupServiceIPs.Insert(lbIP.String())
-			}
-			nodePort := svcInfo.NodePort()
-			if svcInfo.Protocol() == v1.ProtocolUDP && nodePort != 0 {
-				conntrackCleanupServiceNodePorts.Insert(nodePort)
-			}
-		}
-	}
+	start := time.Now()
+	klog.V(4).InfoS("Started to reconcile conntrack entries", "ipFamily", ipFamily)
 
-	klog.V(4).InfoS("Deleting conntrack stale entries for services", "IPs", conntrackCleanupServiceIPs.UnsortedList())
-	for _, svcIP := range conntrackCleanupServiceIPs.UnsortedList() {
-		filters = append(filters, filterForIP(svcIP, v1.ProtocolUDP))
-	}
-	klog.V(4).InfoS("Deleting conntrack stale entries for services", "nodePorts", conntrackCleanupServiceNodePorts.UnsortedList())
-	for _, nodePort := range conntrackCleanupServiceNodePorts.UnsortedList() {
-		filters = append(filters, filterForPort(nodePort, v1.ProtocolUDP))
+	entries, err := ct.ListEntries(ipFamilyMap[ipFamily])
+	if err != nil {
+		klog.ErrorS(err, "Failed to list conntrack entries")
+		return
 	}
 
-	if n, err := ct.ClearEntries(ipFamilyMap[ipFamily], filters...); err != nil {
-		klog.ErrorS(err, "Failed to delete stale service connections")
-	} else {
-		klog.V(4).InfoS("Deleted conntrack stale entries for services", "count", n)
+	// The indices "ipIndex" and "portIndex" represent allowed DNAT flows and are
+	// structured as maps for efficient lookups during cleanup. The key of the maps
+	// is origin (original destination [--orig-dst] for ipIndex and
+	// original port [--org-port-dst] for portIndex) and the value is set of allowed
+	// DNATed destinations (--reply-src).
+
+	// ipIndex maps service IPs (ClusterIP, LoadBalancerIPs and ExternalIPs) with the set of endpointIPs.
+	ipIndex := make(map[string]sets.Set[string])
+	// portIndex maps service NodePort with the set of endpointIPs.
+	portIndex := make(map[int]sets.Set[string])
+
+	for svcName, svc := range svcPortMap {
+		// we are only interested in UDP services
+		if svc.Protocol() != v1.ProtocolUDP {
+			continue
+		}
+
+		endpointIPs := sets.New[string]()
+		for _, endpoint := range endpointsMap[svcName] {
+			// Serving endpoints will never black-hole traffic, we consider all
```

Comment:
That wording is a bit off; the endpoints don't do the black-holing themselves.

I think you want to say "we never want black-hole conntrack entries for serving endpoints"?

Reply:
How about "Conntrack flows for a Service (IP or NodePort) that are DNATed to a serving endpoint will never result in black-holed traffic. All the other DNAT flows for the service will potentially cause traffic black-holing, and we clear those flows during reconciliation."